command/command.go

```diff
@@ -175,70 +177,92 @@ func (w *backgroundCmd) Start() error {
 // KeepAlive will create a goroutine to send regular SIGCONT signal to associated command
 // a single goroutine will be launched no matter how many calls are done to KeepAlive
 func (w *backgroundCmd) KeepAlive() {
+	w.Lock()
+	defer w.Unlock()
+
 	if w.DryRun() {
 		return
 	}
 
-	w.Lock()
-	defer w.Unlock()
-
 	if w.ticker != nil {
 		return
 	}
 
 	w.ticker = time.NewTicker(cmdKeepAliveTickDuration)
 
+	w.keepAliveQuit = make(chan int, 1) // we need a buffered channel so the sender doesn't block
+
 	w.log.Debug("monitoring sending SIGCONT signal to process every 1s")
 
 	go func() {
 		for {
-			if w.ticker == nil {
-				return
-			}
-
-			<-w.safeTicker().C
-
-			proc, err := w.processManager.Find(w.pid)
+			exit, err := w.sendSIGCONTSignal()
 			if err != nil {
-				w.log.Errorw("an error occurred when trying to Find process, stopping to monitor background process, ticker removed", "error", err)
-
-				w.resetTicker()
-
+				w.log.Errorw("an error occurred when sending SIGCONT signal to process, stopping to monitor background process, ticker removed", "error", err)
 				return
 			}
 
-			if err := w.processManager.Signal(proc, syscall.SIGCONT); err != nil {
-				if errors.Is(err, os.ErrProcessDone) {
-					w.log.Infof("process is already finished, skipping sending SIGCONT from now on")
-				} else {
-					w.log.Errorw("an error occurred when sending SIGCONT signal to process, stopping to monitor background process, ticker removed", "error", err)
-				}
-
-				w.resetTicker()
-
+			if exit {
 				return
 			}
 		}
 	}()
 }
 
-func (w *backgroundCmd) safeTicker() *time.Ticker {
+func (w *backgroundCmd) sendSIGCONTSignal() (exit bool, err error) {
 	w.Lock()
 	defer w.Unlock()
 
-	return w.ticker
+	if w.ticker == nil {
+		return true, nil
+	}
+
+	select {
+	case <-w.keepAliveQuit:
+		close(w.keepAliveQuit)
+		w.log.Debug("background process exited, stopping to monitor background process, ticker removed")
+
+		return true, nil
+	case <-w.ticker.C:
+	}
+
+	proc, err := w.processManager.Find(w.pid)
+	if err != nil {
+		w.log.Errorw("an error occurred when trying to Find process, stopping to monitor background process, ticker removed", "error", err)
+
+		w.resetTicker()
+
+		return false, err
+	}
+
+	if err := w.processManager.Signal(proc, syscall.SIGCONT); err != nil {
+		w.resetTicker()
+
+		if errors.Is(err, os.ErrProcessDone) {
+			w.log.Infof("process is already finished, skipping sending SIGCONT from now on")
+
+			return true, nil
+		}
+
+		w.log.Errorw("an error occurred when sending SIGCONT signal to process, stopping to monitor background process, ticker removed", "error", err)
+
+		return false, err
+	}
+
+	return false, nil
 }
 
 func (w *backgroundCmd) resetTicker() {
-	w.Lock()
-	defer w.Unlock()
-
 	if w.ticker == nil {
 		return
 	}
 
 	w.ticker.Stop()
 	w.ticker = nil
+
+	if w.keepAliveQuit != nil {
+		w.keepAliveQuit <- 0
```

Comment:
The issue before was that `keepAliveQuit` was unbuffered, so this would block until someone read from the channel. But when we're called by `Stop()`, who has the lock, we'll never be able to read from this channel in `sendSIGCONTSignal`, as it can't get the lock. If we make it a buffered channel, this won't block anymore. Stop() can complete, and then the goroutine calling sendSIGCONTSignal can complete. I don't see any more stuck pods locally

Reply:
Really interesting 🤔 I learn something today! Thanks for the investigation and the fix! Good job! 💪